ListObjects with prefix and non existing marker causes an NotImplemented error

I am trying to list objects using *github.com/aws/aws-sdk-go/service/s3* package

Pseudo code example:
```
        s3session, err := session.NewSession(&aws.Config{
		Endpoint:         aws.String("http://localhost:9000"),
		Credentials:      credentials.NewStaticCredentials("...", "...", ""),
		Region:           aws.String("us-east-1"),
		DisableSSL:       aws.Bool(true),
		S3ForcePathStyle: aws.Bool(true),
	})
	if err != nil {
		panic(err)
	}

	s3Api := s3.New(s3session)
	resp, err := s3Api.ListObjects(&s3.ListObjectsInput{
		Bucket: aws.String("bucket"),
		Prefix: aws.String("some_prefix"),
		Marker: aws.String("some_marker"),
	})
	if err != nil {
		panic(err)
	}
	fmt.Println(resp)
```

## Expected Behavior
Response is:
```
{
  IsTruncated: false,
  Marker: "some_marker",
  MaxKeys: 1000,
  Name: "bucket",
  Prefix: "some_prefix"
}
```

## Current Behavior
```
panic: NotImplemented: A header you provided implies functionality that is not implemented
	status code: 501, request id: 150A9DB07BBBEB40, host id: 
```

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
1. Run the docker image *docker minio container: docker run --name=s3_server -p 9000:9000 -d minio/minio server /expor*;
2. Create a new bucket via minio web page. Bucket name is "bucket";
3. Execute the code above.

## Context
We use minio for testing, to avoid using real aws s3 service. We are building something like s3 discovery logic, due to the issue our QA were unable to write test cases for a similar situation as I provided above. Using real aws s3 service I was able to perform this request and didn't get any errors.

## Your Environment
* Version used (`minio version`):

```
/ # minio version
Version: 2018-01-02T23:07:00Z
Release-Tag: RELEASE.2018-01-02T23-07-00Z
Commit-ID: a6318dbdaf0a33d80e52bb762e153e4d7ca396f7
```

* Environment name and version (e.g. nginx 1.9.1):

```
❯ docker version
Client:
 Version:       17.12.0-ce
 API version:   1.35
 Go version:    go1.9.2
 Git commit:    c97c6d6
 Built: Wed Dec 27 20:03:51 2017
 OS/Arch:       darwin/amd64

Server:
 Engine:
  Version:      17.12.0-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.2
  Git commit:   c97c6d6
  Built:        Wed Dec 27 20:12:29 2017
  OS/Arch:      linux/amd64
  Experimental: true
```

--- a/cmd/admin-handlers_test.go
+++ b/cmd/admin-handlers_test.go
@@ -696,7 +696,7 @@ func TestValidateHealQueryParams(t *testing.T) {
 			marker:    "notmatchingmarker",
 			delimiter: "/",
 			maxKeys:   "10",
-			apiErr:    ErrNotImplemented,
+			apiErr:    ErrInvalidPrefixMarker,
 		},
 		// 8. Invalid params with unsupported delimiter.
 		{
@@ -812,7 +812,7 @@ func TestListObjectsHealHandler(t *testing.T) {
 			marker:     "notmatchingmarker",
 			delimiter:  "/",
 			maxKeys:    "10",
-			statusCode: getAPIError(ErrNotImplemented).HTTPStatusCode,
+			statusCode: getAPIError(ErrInvalidPrefixMarker).HTTPStatusCode,
 		},
 		// 8. Invalid params with unsupported delimiter.
 		{

--- a/cmd/api-errors.go
+++ b/cmd/api-errors.go
@@ -124,6 +124,7 @@ const (
 	ErrUnsupportedMetadata
 	ErrMaximumExpires
 	ErrSlowDown
+	ErrInvalidPrefixMarker
 	// Add new error codes here.
 
 	// Server-Side-Encryption (with Customer provided key) related API errors.
@@ -524,6 +525,12 @@ var errorCodeResponse = map[APIErrorCode]APIError{
 		Description:    "Please reduce your request",
 		HTTPStatusCode: http.StatusServiceUnavailable,
 	},
+	ErrInvalidPrefixMarker: {
+		Code:           "InvalidPrefixMarker",
+		Description:    "Invalid marker prefix combination",
+		HTTPStatusCode: http.StatusBadRequest,
+	},
+
 	// FIXME: Actual XML error response also contains the header which missed in list of signed header parameters.
 	ErrUnsignedHeaders: {
 		Code:           "AccessDenied",

--- a/cmd/bucket-handlers-listobjects.go
+++ b/cmd/bucket-handlers-listobjects.go
@@ -44,7 +44,7 @@ func validateListObjectsArgs(prefix, marker, delimiter string, maxKeys int) APIE
 	if marker != "" {
 		// Marker not common with prefix is not implemented.
 		if !hasPrefix(marker, prefix) {
-			return ErrNotImplemented
+			return ErrInvalidPrefixMarker
 		}
 	}
 	// Success.
@@ -107,9 +107,18 @@ func (api objectAPIHandlers) ListObjectsV2Handler(w http.ResponseWriter, r *http
 	// Validate the query params before beginning to serve the request.
 	// fetch-owner is not validated since it is a boolean
 	if s3Error := validateListObjectsArgs(prefix, marker, delimiter, maxKeys); s3Error != ErrNone {
+		// return empty response if invalid marker
+		//TODO: avoid this pattern when moving to errors package
+		if s3Error == ErrInvalidPrefixMarker {
+			listObjectsInfo := ListObjectsInfo{}
+			response := generateListObjectsV2Response(bucket, prefix, token, marker, startAfter, delimiter, fetchOwner, listObjectsInfo.IsTruncated, maxKeys, listObjectsInfo.Objects, listObjectsInfo.Prefixes)
+			writeSuccessResponseXML(w, encodeResponse(response))
+			return
+		}
 		writeErrorResponse(w, s3Error, r.URL)
 		return
 	}
+
 	// Inititate a list objects operation based on the input params.
 	// On success would return back ListObjectsInfo object to be
 	// marshalled into S3 compatible XML header.
@@ -151,6 +160,13 @@ func (api objectAPIHandlers) ListObjectsV1Handler(w http.ResponseWriter, r *http
 
 	// Validate all the query params before beginning to serve the request.
 	if s3Error := validateListObjectsArgs(prefix, marker, delimiter, maxKeys); s3Error != ErrNone {
+		// return empty response if invalid marker
+		//TODO: avoid this pattern when moving to errors package
+		if s3Error == ErrInvalidPrefixMarker {
+			response := generateListObjectsV1Response(bucket, prefix, marker, delimiter, maxKeys, ListObjectsInfo{})
+			writeSuccessResponseXML(w, encodeResponse(response))
+			return
+		}
 		writeErrorResponse(w, s3Error, r.URL)
 		return
 	}

